dtm0/fop.c

```diff
@@ -341,50 +349,92 @@ M0_INTERNAL void m0_dtm0_on_committed(struct m0_fom                *fom,
 
 static int dtm0_fom_tick(struct m0_fom *fom)
 {
-	int                     rc;
-	struct dtm0_req_fop    *req;
-	struct dtm0_rep_fop    *rep;
-	struct m0_dtm0_service *svc;
-
-	if (m0_fom_phase(fom) < M0_FOPH_NR) {
+    int                       rc;
+    struct   dtm0_req_fop    *req = m0_fop_data(fom->fo_fop);
+    struct   dtm0_rep_fop    *rep;
+    struct   m0_dtm0_service *svc;
+    uint64_t                  phase_sm_id;
+    uint64_t                  rpc_sm_id;
+    struct   m0_be_tx_credit  dtm0logrec_cred = {};
+    struct   m0_dtm0_tx_desc  txd = {};
+    struct   m0_buf           buf = {};
+    struct   m0_rpc_item     *item = &fom->fo_fop->f_item;
+    int                       phase = m0_fom_phase(fom);
+
+    M0_ENTRY("fom %p phase %d", fom, phase);
+    switch (phase) {
+    case M0_FOPH_INIT ... M0_FOPH_NR - 1:
 		rc = m0_fom_tick_generic(fom);
-	} else {
-		M0_ASSERT(m0_fom_phase(fom) == M0_FOPH_TYPE_SPECIFIC);
-		req = m0_fop_data(fom->fo_fop);
-		rep = m0_fop_data(fom->fo_rep_fop);
-		M0_SET0(&rep->dr_txr);
-		svc = m0_dtm0_service_find(fom->fo_service->rs_reqh);
-		M0_ASSERT(svc != NULL);
-
-		M0_LOG(M0_DEBUG, "Processing non-generic phase %" PRIu32
-		       " %d, " FID_F ", %p", req->dtr_msg,
-		       !!m0_dtm0_is_a_volatile_dtm(fom->fo_service),
-		       FID_P(&fom->fo_service->rs_service_fid),
-		       fom->fo_service->rs_reqh);
-
-		if (m0_dtm0_in_ut() && req->dtr_msg == DMT_EXECUTE &&
-		    m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
-			rc = m0_dtm0_tx_desc_copy(&req->dtr_txr,
-						  &rep->dr_txr);
-			M0_ASSERT(rc == 0);
-			m0_dtm0_send_msg(fom, DMT_EXECUTED,
-					 &req->dtr_txr.dtd_id.dti_fid,
-					 &req->dtr_txr);
-		}
+		break;
 
+	case M0_FOPH_DTM0_PREP:
 		if (req->dtr_msg == DTM_PERSISTENT &&
-		    m0_dtm0_is_a_volatile_dtm(fom->fo_service)) {
-			/* TODO: Only the client side is here so far.
-			 * Remove the "is_volatile" part when plog is ready.
-			 */
-			m0_be_dtm0_log_pmsg_post(svc->dos_log, fom->fo_fop);
+			m0_dtm0_is_a_persistent_dtm(fom->fo_service)) {
+			rc = m0_dtm0_tx_desc_copy(&req->dtr_txr, &txd);
+			if (rc != 0) {
+				m0_fom_phase_move(fom, M0_ERR(rc), M0_FOPH_FAILURE);
+			}
+
+			m0_be_dtm0_log_credit(M0_DTML_PERSISTENT,
+								  &txd,
+								  &buf,
+								  m0_fom_reqh(fom)->rh_beseg,
+								  NULL,
+								  &dtm0logrec_cred);
+			m0_be_tx_credit_add(&fom->fo_tx.tx_betx_cred, &dtm0logrec_cred);
+			rc = m0_be_tx_open_sync(&fom->fo_tx.tx_betx);
```

Comment:
Why do we need to explicitly open_sync/close_sync this tx?  Wouldn't it be done as a part of the generic phases?

Example:

```
case <generic-phases>:
    rc = tick_generic(fom);
    next_phase = fom_phase(fom);
    /* Calculate the credit before opening the tx */
    if (next_phase == TX_OPEN)
        caclulate_log_record_credits(...);
    break;
```

Reply:
Applied the changes as discussed.